File: clojure/problem-4.clj
; A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.
; Find the largest palindrome made from the product of two 3-digit numbers.

(defn palindrome-number? [num]
  (let [numstr (str num)
        mid    (/ (count numstr) 2)
        str1   (subs numstr 0 mid)
        str2   (clojure.string/reverse (subs numstr mid))]
    (= str1 str2)))

(let [start 100
      end 999]
  (for [x (range start end)
        :let [y (* x x)]
        :when (palindrome-number? y)]
    y))

File: go/problem4.go
// A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

// Find the largest palindrome made from the product of two 3-digit numbers.

package main

import "fmt"

func is_palindrome(num int) bool {
  rev, tmp := 0, num
  for ; tmp > 0; {
    rev = (rev * 10) + (tmp % 10)
    tmp = tmp / 10
  }
  return num == rev
}

func main(){
  max, p := 0, 0

  for i := 1; i < 100; i++ {
    for j := 1; j < 100; j++ {
      p = i * j
      if is_palindrome(p) {
        max = p
      }
    }
  }

  fmt.Println(max)
}